rfcs/convergence/make-overrides.md

```diff
@@ -0,0 +1,195 @@
+
+To solve these issues we made a step back to the original API of `makeStyles()` (current iteration is a second version) and as we introduced `ax()` to merge atomic classnames we explored different solutions.
+
+Proposed API solve the problem in a "vandal" way 🪓 Matchers are moved to user's scope thus we can have all definitions in a single `makeStyles()` call => we have a single React hook. See a modified snippet below:
+
+```tsx
+import { ax, makeOverrides } from "@fluentui/react-make-styles";
+
+const useStyles = makeStyles({
+  /* 👍 no matchers, no need to execute on each render */
+  root: { color: "red" },
+  rootPrimary: { color: "blue" },
+
+  /* 👍 styles for each slot are defined together (not a requirement) */
+  icon: { background: "black" },
+  iconPrimary: { background: "white" }
+});
+
+function Component() {
+  /* 👍 a single call of React hook */
+  const classes = useStyles();
+
+  return (
+    <div
+      className={ax(
+        classes.root /* The concept of matching is replaced with selective classname concat */,
+        props.primary && classes.rootPrimary,
+        props.className
+      )}
+    >
+      {props.icon && (
+        <div
+          className={ax(
+            classes.icon,
+            props.primary && classes.iconPrimary,
+            props.icon.className
+          )}
+        />
+      )}
+    </div>
+  );
+}
+```
+
+### Pros and Cons
+
+These changes result in performance improvements in `makeOveriddes`:
+
+- matcher functions are no longer executed on each render in styling functions
+- styles are executed and merged only when slots are present
+- less React hooks to call for slots
+
+Surprisingly, this also moves us closer to [CSS Modules](https://github.com/css-modules/css-modules). The same snippet with CSS Modules, for example:
+
+```tsx
+import cx from "classnames";
+import * as classes from "./Component.css"; // the same styles can be written in CSS
+
+function Component() {
+  return (
+    <div
+      className={cx(
+        classes.root,
+        props.primary && classes.rootPrimary,
+        props.className
+      )}
+    >
+      {props.icon && (
+        <div
+          className={cx(
+            classes.icon,
+            props.primary && classes.iconPrimary,
+            props.icon.className
+          )}
+        />
+      )}
+    </div>
+  );
+}
+```
+
+In [microsoft/fluentui#16923](https://github.com/microsoft/fluentui/pull/16923)  `Avatar`, from `@fluentui/react-avatar` (check `packages/react-avatar/src/components/Avatar/useAvatarStyles.ts`), was used to validate performance improvements and explore potential issues. It is uncertain if/how these issues are critical, but there are two things that should be mentioned:
+
+```tsx
+const useStyles = makeOverrides({
+  /* ... Complex apps/components can contain many definitions ... */
+
+  rootShape20: { width: '20px', height: '20px' },
+  rootShape24: { width: '24px', height: '24px' },
+  rootShape28: { width: '28px', height: '28px' },
+})
```

Comment:
I do quite like this pattern of defining styles ahead of time, then conditionally applying them below. We should measure, but my intuition says that this will have good opportunity for optimization and gets us closer to static styles. 

I also find this more intuitive than the matcher approach 👍  I think this will be easier for devs to understand and other teams to use in their products.

Reply:
> I also find this more intuitive than the matcher approach 👍 I think this will be easier for devs to understand and other teams to use in their products.

There is one interesting difference:
- with matchers (i.e. `makeStyles()`) you can just execute a function and get classes back
- with `makeOverrides()` you need to execute component's code to understand applied styles

I am considering this an implementation difference that should not block us from moving forward.

Reply:
> Question: What prevents us from abstracting the component code into a separate function so that we can, like you said, just execute a function and get classes back?

```tsx
function getRootClasses(state, classes) {
  return ax(classes.root, state.primary && classes.primary, /* etc. */)
}
```

Do you mean something like this?